packages/workspace/src/browser/workspace-trust-service.ts

```diff
@@ -0,0 +1,115 @@
+/********************************************************************************
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { Dialog, StorageService } from '@theia/core/lib/browser';
+import { FrontendApplicationConfigProvider } from '@theia/core/lib/browser/frontend-application-config-provider';
+import { PreferenceService } from '@theia/core/lib/browser/preferences/preference-service';
+import { MessageService } from '@theia/core/lib/common/message-service';
+import { nls } from '@theia/core/lib/common/nls';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { inject, injectable, postConstruct } from '@theia/core/shared/inversify';
+import { WorkspaceService } from '.';
+import {
+    WorkspaceTrustPreferences, WORKSPACE_TRUST_EMPTY_WINDOW, WORKSPACE_TRUST_ENABLED, WORKSPACE_TRUST_STARTUP_PROMPT, WORKSPACE_TRUST_STARTUP_PROMPT_NEVER,
+    WORKSPACE_TRUST_STARTUP_PROMPT_ONCE
+} from './workspace-trust-preferences';
+
+const CONFIG_AUTO_TRUST = 'autoTrust';
+const STORAGE_TRUSTED = 'trusted';
+
+@injectable()
+export class WorkspaceTrustService {
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(PreferenceService)
+    protected readonly preferences: PreferenceService;
+
+    @inject(StorageService)
+    protected readonly storage: StorageService;
+
+    @inject(MessageService)
+    protected readonly messageService: MessageService;
+
+    @inject(WorkspaceTrustPreferences)
+    protected readonly workspaceTrustPref: WorkspaceTrustPreferences;
+
+    protected trusted = new Deferred<boolean>();
+    protected _isTrusted: boolean | undefined;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        await this.workspaceService.ready;
+        this.trusted.promise.then(trusted => this._isTrusted = trusted);
+        const implicitTrust = await this.calculateWorkspaceTrust();
+        if (implicitTrust !== undefined) {
+            this.trusted.resolve(implicitTrust);
+        }
```

Comment:
Does this need to be run when this class is initialized, or could we wait until the first call to `requestWorkspaceTrust`?

Reply:
Ideally yes, since in the plugin API the 'onDidGrantWorkspaceTrust' which we trigger after resolving this sets the trusted flag on the workspace (see plugin-context.ts and workspace.ts in plugin-ext). The Git plugin, for instances, has code like this:
```
if (workspace.isTrusted) {
   this.disposables.push(new AutoFetcher(this, globalState));
} else {
   const trustDisposable = workspace.onDidGrantWorkspaceTrust(() => {
      trustDisposable.dispose();
      this.disposables.push(new AutoFetcher(this, globalState));
   });
   this.disposables.push(trustDisposable);
}
```
So they query the flag early on and afterwards listen to any potential changes. Therefore we should set and propagate that flag as early as possible if we already know what the value will be. What do you think?

Reply:
Sounds fair to me - I'm not very familiar with the plugin side of things, but if they're expecting that it might be resolved before they're activated, then we should make the effort to satisfy that expectation. 👍